### PR TITLE
transform_feature_names for decomposition and scaling

### PR DESCRIPTION
**Pull request by jnothman**
Present top linear weights as `PCA2:=(x0*5.3+x1*77)` etc.

Issues:
* others may feel this naming convention is too verbose. Suggestions welcome
* `make_tfn_weighted` should probably have a doctest (but I'm not sure how to do this without modifying the global register of `transform_feature_names` implementations)
* tests may be too slow, particularly those using `hypothesis`. Suggestions welcome.

### CODE REVIEW

**Comment by jnothman** [comment]
In merging with master, I see that there has been an attempt to more thoroughly test every estimator with `transform_feature_names`. Where multiple are registered identically, I believe this is akin to testing every value in a for loop and is excessive. I would rather leave the tests readable.

**Comment by kmike** [comment]
Hey @jnothman! I'll check this PR later today. 

As for excessive testing, I like testing of all (most) estimators explicitly because of two reasons:

1. this way we declare explicitly what's supported and what's not supported;
2. it helps to find bugs and wrong assumptions. 

For example, RandomizedLogisticRegression has SelectorMixin as a metaclass, not as a base class, so https://github.com/TeamHG-Memex/eli5/pull/177 didn't work for it; I found this when adding all these estimators (which I was thinking are supported) one-by-one. Another example is lightning support: when adding all estimators on-by-one I found several inconsistencies which required fixing in lightning: https://github.com/scikit-learn-contrib/lightning/pull/95, https://github.com/scikit-learn-contrib/lightning/pull/98.

I can see how from a library author (sklearn, etc.) point of view these tests look unnecessary. But they require knowledge of inner workings of the library, and knowledge of development plans - e.g. SelectorMixin is not documented, it is not a part of public scikit-learn API; as a scikit-learn core developer you are comfortable relying on it, but as an outsider I like to be more explicit in what eli5 supports, and check one more time that the library interface is consistent and there are no gotchas and hidden issues. In case of #177 it was also a learning excercise for me - I was checking what this PR means for the user.

**Comment by jnothman** [comment]
The issue with randomized l1 is not that SelectorMixin is a metaclass; it
is not. It's that SelectorMixin was only made a base class of it recently
(i.e. not in a released version).

You're probably right. But in testing everything, you will also need to
manage deprecations and newcomers in order to support multiple versions. It
will be maintenance hell in time...

**Comment by kmike** [comment]
A good catch about randomized l1! 

Yeah, deprecations can be a problem. We "solved" it once by dropping sklearn 0.17 support, and likely it'll cause problems again with 0.19. But to my taste it is a better problem to have: it is testing-only, and it is visible; if we don't test then problem can be in user applications, and it can be undetected by our testing suite. But yeah, explicit estimator support increases effort required for maintenance.

**Comment by jnothman** [comment]
I've not worked out how to fix `mypy`'s complaint that
```
eli5/sklearn/transform.py:176: error: Argument 1 has incompatible type "make_tfn_weighted"; expected Callable[..., Any]
```

**Comment by kmike** [comment]
@jnothman yep, mypy still has a lot of rough edges, even for the basics - https://github.com/python/mypy/issues/797. You can ignore such errors like this:

```py
transform_feature_names.register(TfidfTransformer)(  # type: ignore
    make_tfn_weighted('idf_', func_name='TFIDF', show_idx=False))
```

**Comment by kmike** [comment]
Hey @jnothman,

I've tried it with LSI: 

```
vec = CountVectorizer()
tfidf = TfidfTransformer()
svd = TruncatedSVD(n_components=50)
clf = LogisticRegressionCV()
pipe = make_pipeline(vec, tfidf, svd, clf)
```

eli5.explain_weigths shows features like this now:

```
SVD2:=(TFIDF(keith*3.75)*0.362+TFIDF(caltech*3.99)*0.27+TFIDF(livesey*4.55)*0.223)
```

Unlike TfidfVectorizer, it shows idf weights, which is nice; I wonder if we should use idf_ as `coef_scale` by default for TfidfVectorizer.

What I found confusing is `TFIDF(keith*3.75)` notation: it reads like "take a number of words "keith" in document, multiply by 3.75, then compute TF*IDF for it (?)", while in fact IDF is 3.75, so it is `TF(keith)*3.75`. If the goal is not to show a math formula, but to show transformer name, then what do you think about using some other notation, e.g. `TFIDF[...]` or something else. Ideas?

**Comment by jnothman** [comment]
I agree that it looks like function application, which is wrong. Square or
other brackets help a little. Not sure what alternatives are

On 29 May 2017 11:59 pm, "Mikhail Korobov" <notifications@github.com> wrote:

Hey @jnothman <https://github.com/jnothman>,

I've tried it with LSI:

vec = CountVectorizer()
tfidf = TfidfTransformer()
svd = TruncatedSVD(n_components=50)
clf = LogisticRegressionCV()
pipe = make_pipeline(vec, tfidf, svd, clf)

eli5.explain_weigths shows features like this now:

SVD2:=(TFIDF(keith*3.75)*0.362+TFIDF(caltech*3.99)*0.27+TFIDF(livesey*4.55)*0.223)

Unlike TfidfVectorizer, it shows idf weights, which is nice; I wonder if we
should use idf_ as coef_scale by default for TfidfVectorizer.

What I found confusing is TFIDF(keith*3.75) notation: it reads like "take a
number of words "keith" in document, multiply by 3.75, then compute TF*IDF
for it (?)", while in fact IDF is 3.75, so it is TF(keith)*3.75. If the
goal is not to show a math formula, but to show transformer name, then what
do you think about using some other notation, e.g. TFIDF[...] or something
else. Ideas?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/TeamHG-Memex/eli5/pull/208#issuecomment-304667660>, or mute
the thread

**Comment by kmike** [comment]
FTR: tests are failing because of https://github.com/TeamHG-Memex/eli5/issues/217.

**Comment by jnothman** [comment]
I'd really like to hear some feedback on this, as I feel linear transformation is a very common case, but it is admittedly hard to present nicely as a string.

**Comment by jnothman** [comment]
Otoh if you feel like it's stretching the scope of eli5 too wide, I'm happy to make a separate project of it.

**Comment by kmike** [comment]
@jnothman this PR looks great, I'm looking forward to using it myself. My only reservation is `tf(...)` which looks like a function is applied. It seems the easiest way to fix it is to make formatting in `make_tfn_weighted` customizable, and customize it for TfidfTransformer.

**Comment by jnothman** [comment]
well maybe that bracketing is wrong in all the cases. something less loaded
like angle brackets?

**Comment by kmike** [comment]
@jnothman yes, sounds fine.

I tried `<>` as well as few other options (`{}`, `[]`), they all have disadvantages - `{}` looks like a set notation, `[]` looks like we're showing a vector, `<>` may be confused with greater/less signs. But `<>` looks less ambiguous, it is no worse than alternatives:

```
SVD2:=<TFIDF<keith*3.75>*0.362+TFIDF<caltech*3.99>*0.27+TFIDF<livesey*4.55>*0.223>
SVD2:={TFIDF{keith*3.75}*0.362+TFIDF{caltech*3.99}*0.27+TFIDF{livesey*4.55}*0.223}
SVD2:=[TFIDF[keith*3.75]*0.362+TFIDF[caltech*3.99]*0.27+TFIDF[livesey*4.55]*0.223]
SVD2:=(TFIDF(keith*3.75)*0.362+TFIDF(caltech*3.99)*0.27+TFIDF(livesey*4.55)*0.223)

PCA2:={x0*5.3+x1*77}
PCA2:=<x0*5.3+x1*77>
PCA2:=[x0*5.3+x1*77]
PCA2:=(x0*5.3+x1*77)
```

Maybe spaces around `+` can help a bit with readability, what do you think?

```
SVD2:=<TFIDF<keith*3.75>*0.362 + TFIDF<caltech*3.99>*0.27 + TFIDF<livesey*4.55>*0.223>
PCA2:=<x0*5.3 + x1*77>
```

**Comment by kmike** [comment]
Does it make sense to keep `()` for `:=` notation?

```
SVD2:=(TFIDF<keith*3.75>*0.362 + TFIDF<caltech*3.99>*0.27 + TFIDF<livesey*4.55>*0.223)
PCA2:=(x0*5.3 + x1*77)
```

**Comment by jnothman** [comment]
You're right < is going to break when we support discretizing.

**Comment by kmike** [comment]
What do you think about the following?

1) add spaces around `+` signs;
2) add '...' if there are missing terms in a sum, e.g. `PCA2:=(x0*5.3 + x1*77 + ...)`;
3) keep `()` braces for everything except for tf*idf;
4) change tf*idf formatting. Options: `TFIDF/keith*3.75/`, `TF(keith)*3.75`, `TF<keith*3.5>`, `TFIDF[keith*3.75]`, `TFIDF:(keith*3.75)`.

(4) is what I'm less certain in. Maybe we should just pick one of the options even if it is not perfect to move it forward; it can be fixed later if someone comes with a better idea.

**Comment by jnothman** [comment]
`TF(keith)*3.75` would make sense... But maybe it belongs in a different function and I've loaded too much on here.

**Comment by jnothman** [comment]
With the `...` idea, do you want to see this ellipsis only when `top` is the reason for limiting results, or even when `threshold` is limiting results?

**Comment by jnothman** [comment]
I've not put in ellipsis previously because I considered brevity to be a virtue in feature names.

**Comment by kmike** [comment]
@jnothman Hm, I was thinking about both top and threshold. We shouldn't show "+ …" if the rest of features are zero though, and if you think of threshold as of a way to filter out near-zero components I see how it can be different from `top`.

`PCA2:=(x0*0.325 + x1*0.221 + …)` is a bit longer, but e.g. for 

`SVD2:=(TFIDF(keith)*3.75*0.362 + TFIDF(caltech)*3.99*0.27 + TFIDF(livesey)*4.55*0.223 + …)`

it doesn't matter. 

It bothers me a bit that these are displaying options, so they'd be better handled at formatting level. For example, the rest of terms in SVD can be displayed on mouse hover in HTML. But it looks outside the scope of this pull request; it may require a structured representation of feature names, instead of plain strings everywhere. 

We had this issue with HashingVectorizer, where we needed to store signs of possible feature names, in addition to names themselves; that's why formatters accept feature names which are lists of {'name': name, 'sign': sign} dicts.
